mirgecom/euler.py

```diff
@@ -139,29 +146,55 @@ def _aux_shape(ary, leading_shape):
         return ()
 
 
+def get_num_species(dim, q):
+    """Return number of mixture species."""
+    return len(q) - (dim + 2)
+
+
+def get_num_conserved(dim, q):
+    """Return number of conserved quantities."""
+    return dim + 2 + get_num_species(dim, q)
+
+
 def split_conserved(dim, q):
     """Get the canonical conserved quantities.
 
     Return a :class:`ConservedVars` that is the canonical conserved quantities,
     mass, energy, and momentum from the agglomerated object array extracted
-    from the state vector *q*.
+    from the state vector *q*. If present, species mass fractions are included.
     """
-    assert len(q) == 2+dim
-    return ConservedVars(mass=q[0], energy=q[1], momentum=q[2:2+dim])
+    #    assert len(q) == dim + 2 + get_num_species(dim, q)
+    nspec = get_num_species(dim, q)
+    if nspec > 0:
+        return ConservedVars(mass=q[0], energy=q[1], momentum=q[2:2+dim],
+                             massfractions=q[2+dim:2+dim+nspec])
+    else:
+        return ConservedVars(mass=q[0], energy=q[1], momentum=q[2:2+dim])
```

Comment:
I hadn't really thought about this, but it should be spelled out in the documentation (perhaps of `mirgecom.euler`): I'm guessing the mass fractions add up to one. If so, do we need all n or would n-1 suffice? (and which would should be the "1"?) Can we make it so that there are no branches and the same logic applies always? (I.e. loops over mass fractions transparently become no-ops.) One thing that might help might be to make the mass fractions an empty object array.

Reply:
For just plain-ole scalar advection - which is what we are adding here - there is no such restriction on the Ys.  In fact, they are completely independent.  When we add mixture EOS and mixture Initializers, the restriction of SUM(Y) = 1 will be introduced, and we are currently discussing whether this will be a restriction we need to enforce or whether it will be one we simply monitor.

Reply:
> we are currently discussing whether this will be a restriction we need to enforce or whether it will be one we simply monitor.

Where is that discussion taking place?

> When we add mixture EOS and mixture Initializers, the restriction of SUM(Y) = 1 will be introduced

Does it make sense to write code here without knowing the answer to the question above (given what this code should look like depends on the answer to that question)?

Reply:
There are no such requirements on these general transported scalars. They do not need to sum to 1, or have any other properties.  Any such properties need to be imposed by the EOS.  It does not make sense to introduce this arbitrary restriction on the scalars at this point.

Reply:
> It does not make sense to introduce this arbitrary restriction on the scalars at this point.

Fair enough. In that case, `nspec` should be renamed `nscalars`.

Reply:
I agree completely.  It might be advantageous to allow this slight of expression, at least temporarily.  Can we talk about this again after the full stack merge with Pyrometheus EOS and everything?

Reply:
It doesn't make me happy, but sure: #165

Also, please get out of the habit of writing things like `nspec` when `nspecies` takes the same time to type. Why sacrifice clarity for new users?